### DISCUSSION TITLE
How do you pass around query results between Python functions

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#10962] [Q&A]
I'm working on a project where I want to use duckdb merely for its analytical capabilities and not for persistence.
As such, I'm trying to pass around result sets between functions, as if I was working with a dataframe library.

I have found this very perplexing and have previously [opened a ticket](https://github.com/duckdb/duckdb/issues/10954) about what I thought was a bug but I believe turned out to be correct behavior.

Nevertheless, I still haven't managed to achieve what I wanted, and I think it's essentially due to the fact that the duckdb Python API doesn't *really* have the concept of a results set object.

I have boiled down my problems of working around this into three MWE, which together seem to make it impossible to do what I'm trying to achieve (whether there are bugs or not). If anybody could show me how I could set things up sanely, that'd be very much appreciated.

Note: I *could* use pandas as intermediate representation, but that (a) is inefficient (b) makes me need to use pandas in places I wouldn't want to have to use pandas (c) exposes me to datatype conversion problems (of which I personally have already reported #9939  #10155  #10196  #10221 -- most get fixed impressively quickly, but some are blocked on pandas, and in general I just rather wouldn't.)

### First MWE: Using a local connection and no third-party dataframes.
(Note: I wrapped each MWE in a function so if you're using Jupyter, your global namespace doesn't get polluted and you don't lose reproducibility if you play around with these.)

```python
def mwe1():
    import duckdb
    def f():
        conn = duckdb.connect(':memory:')
        c = conn.sql("SELECT * FROM (VALUES (0))")
        #str(c)
        conn.close()
        return c
    print(f())
mwe1()
```

The above throws a ConnectionException. If I uncomment the `str(c)`, it prints the "correct" result, but it's inefficient to create throwaway string representation of all my result sets; and the third MWE below shows that this isn't actually  always a solution.

### Second MWE: Using the global/static/shared connection and local pandas dataframes

```python
def mwe2():
    import pandas as pd
    import duckdb
    def f():
        df = pd.DataFrame({'a': [0]})
        c = duckdb.sql("SELECT * FROM df")
        #str(c) 
        return c
    #df = pd.DataFrame({'a': [0, 1, 2]})
    print(f())
mwe2()
```
The above throws a CatalogError. 

(a) If I uncomment only the `str(c)` it prints the "correct" result 
(b) If I uncomment  only the `df = ` it prints the "wrong" result.
(c) If I uncomment both lines, it prints the "correct" result again.

### Third MWE: Using a local connection and local pandas dataframes

```python
def mwe3():
    import duckdb
    import pandas as pd
    def f():
        conn = duckdb.connect(':memory:')
        df = pd.DataFrame({'a': [0, 1, 2]})
        c = conn.sql("SELECT * FROM df WHERE a > 0")
        #str(c)
        conn.close()
        return c
    res = f()
    #df = pd.DataFrame({'a': [0, 1, 2, 3]})
    print(res)
    print(duckdb.query('SELECT * FROM res WHERE a > 1'))
mwe3()
```
Both print statements in the above throw errors, the first a ConnectionException, the second a CatalogError.
(a) If I uncomment only the `str(c)`, the first print statement gives the "correct" result and the second still errors.
(b) If I uncomment only the `df = `, the first print statement still errors and the second gives the "wrong" result
(c) If I uncomment both lines, the first print statement gives the "correct" result and the second the "wrong" result.

**Comment by Tishj** [comment]
First MWE:
(1.a) The `__str__` method of the Relation will cache, so what you're seeing is a cached stringification.

Second MWE:
(2.a) see 1.a
(2.b) Relations are lazy, you're not executing the Relation with the first version of `df`
(2.c) see 1.a, it's cached so it doesn't use the new version of `df`, the result is already computed.

Third MWE:
(3.a) see 1.a
(3.b)  The relation is linked to the connection, if you explicitly close the connection it can't execute anymore
I have to admit I'm not sure why it would throw a CatalogError, what is the full error message there?
I assume it just can't find a suitable replacement scan in `res` so it errors
(3.c) see 1.a, the second probably gives a "wrong" result by showing the result executed with the second version of `df`?

**Reply by soerenwolfers** [reply]
(1.a): I don't think it's a stringification (or I don't understand what you mean by that). I can perform operations on it like on any other table
```
import duckdb
def mwe1():
    import duckdb
    def f():
        conn = duckdb.connect(':memory:')
        c = conn.sql("SELECT * FROM (VALUES (0), (1))")
        str(c)
        conn.close()
        return c
    a = f()
    print(duckdb.query('SELECT * FROM a WHERE col0 > 0'))
mwe1()
```
prints
```
┌───────┐
│ col0  │
│ int32 │
├───────┤
│     1 │
└───────┘
```

**Reply by soerenwolfers** [reply]
Oh, but this also works if I drop the `str(c)`. So the output of `conn.sql` stays alive somehow, I just need to use it from another connection rather than just print it?

**Reply by soerenwolfers** [reply]
3.b: This throws a ConnectionError, not a CatalogError

**Reply by soerenwolfers** [reply]
3.c: Yes, uses the second version of df

**Reply by Tishj** [reply]
> 3.b: This throws a ConnectionError, not a CatalogError

You mentioned both, I addressed the connection error, I'm a little confused by the catalog error

**Reply by Tishj** [reply]
> (1.a): I don't think it's a stringification

It might just be a forced execution then, in which case I think it can be substituted with `c.execute()` instead?

**Reply by soerenwolfers** [reply]
Ah, so the original MWE3 without any modifications except commenting out the first print, throws 
```
CatalogException: Catalog Error: Table with name df does not exist!
Did you mean "pg_am"?
```
 on the second print

**Reply by soerenwolfers** [reply]
With all this being said, I am now at the following status of understanding:

- MWE1: The result object from my example code is "alive and well", but it is lazy; I just have to put it through a connection if I want to see it printed. However, I would prefer being able to have execution happen at the time of the function call.
- MWE2: I can replace `sql` by `execute` to solve the problem of the unavailable or aliased local dataframe. However, that means I'm passing around a `connection` object, which cannot be used in replacement scans, and the only way to get a result set out of it seems to be `.df()` which I rather wouldn't. Also, even worse, since I'm using a global connection here, the "pending results set" on that connection would be overwritten by another function call.
- MWE3: I can't use `execute`  to begin with, because I'm closing the connection within the function call.

**Reply by Tishj** [reply]
That's not what I said, I said use the `.execute()` method of the Relation
(I know, slightly confusing)

`duckdb.sql(....).execute()` will execute the relation

**Reply by soerenwolfers** [reply]
That still gives me a CatalogError:
```
def mwe2():
    import pandas as pd
    import duckdb
    def f():
        df = pd.DataFrame({'a': [0]})
        c = duckdb.sql("SELECT * FROM df").execute()
        #str(c) 
        return c
    #df = pd.DataFrame({'a': [0, 1, 2]})
    print(f())
mwe2()
```
```
InvalidInputException: Invalid Input Error: Attempting to execute an unsuccessful or closed pending query result
Error: Catalog Error: Table with name df does not exist!
Did you mean "pg_am"?
```

**Reply by soerenwolfers** [reply]
(`sql(...).execute()` does solve my problems with MWE1 though, thanks! Unfortunately, in most cases, I do use some variables from the local function context)

**Comment by soerenwolfers** [comment]
Sounds like my problems here will be solved by a resolution of https://github.com/duckdb/duckdb/issues/11224
